src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -401,6 +401,79 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                     unreached();
             }
         }
+        else if (intrin.numOperands >= 2 && intrin.op2->IsEmbMaskOp())
+        {
+            // Handle case where op2 is operation that needs embedded mask
+            GenTree* op2 = intrin.op2;
+            assert(intrin.id == NI_Sve_ConditionalSelect);
+            assert(op2->isContained());
+            assert(op2->OperIsHWIntrinsic());
+
+            // Get the registers and intrinsics that needs embedded mask
+            const HWIntrinsic intrinOp2(op2->AsHWIntrinsic());
+            instruction       insOp2     = HWIntrinsicInfo::lookupIns(intrinOp2.id, intrinOp2.baseType);
+            const bool        instrIsRMW = op2->isRMWHWIntrinsic(compiler);
+
+            regNumber maskReg     = op1Reg;
+            regNumber instrOp1Reg = REG_NA;
+            regNumber instrOp2Reg = REG_NA;
+            regNumber falseReg    = op3Reg;
+
+            switch (intrinOp2.numOperands)
+            {
+                case 2:
+                    assert(intrinOp2.op2 != nullptr);
+                    instrOp2Reg = intrinOp2.op2->GetRegNum();
+                    FALLTHROUGH;
+
+                case 1:
+                    assert(intrinOp2.op1 != nullptr);
+                    instrOp1Reg = intrinOp2.op1->GetRegNum();
+                    break;
+
+                default:
+                    unreached();
+            }
+
+            switch (intrinOp2.numOperands)
+            {
+                case 1:
+                    GetEmitter()->emitIns_R_R_R(insOp2, emitSize, targetReg, maskReg, instrOp1Reg, opt);
+                    break;
+
+                case 2:
+
+                    assert(instrIsRMW);
+                    if (targetReg != falseReg)
+                    {
+                        assert(targetReg != instrOp1Reg);
+                        assert(targetReg != instrOp2Reg);
+
+                        if (!intrin.op3->IsVectorZero())
+                        {
+                            // If the `falseValue` of conditional-select is non-zero, then we start with that in the
+                            // destination followed by `movprfx` (see below) of the relevant bits of the first operand
+                            // of `insOp2` based upon the value of predicate `cond`.
+                            GetEmitter()->emitIns_Mov(INS_mov, emitTypeSize(node), targetReg, falseReg,
```

Comment:
Have we checked yet if this can be `movprfx (unpredicated)` the docs indicate
> Although the operation of the instruction is defined as a simple predicated vector copy, it is required that the prefixed instruction at PC+4 must be an SVE destructive binary or ternary instruction encoding, or a unary operation with merging predication, but excluding other MOVPRFX instructions. The prefixed instruction must specify the same predicate register, and have the same maximum element size (ignoring a fixed 64-bit "wide vector" operand), and the same destination vector as the MOVPRFX instruction. The prefixed instruction must not use the destination register in any other operand position, even if they have different names but refer to the same architectural register state. Any other use is UNPREDICTABLE.

and it isn't exactly clear if that allows or disallows multiple `movprfx` sequential instructions. Specifically the ambiguity comes from the following, since it says required to be but excluding other movprfx, and that isn't 100% clear to me:
> it is required that the prefixed instruction at PC+4 must be an SVE destructive binary or ternary instruction encoding, or a unary operation with merging predication, but excluding other MOVPRFX instructions

Reply:
yes, that's something that I will bring up with @a74nh tomorrow.

Reply:
> and it isn't exactly clear if that allows or disallows multiple `movprfx` sequential instructions. Specifically the ambiguity comes from the following, since it says required to be but excluding other movprfx, and that isn't 100% clear to me:
> 
> > it is required that the prefixed instruction at PC+4 must be an SVE destructive binary or ternary instruction encoding, or a unary operation with merging predication, but excluding other MOVPRFX instructions

The instruction after a `movprfx` can't be another `movprfx`.

@TamarChristinaArm just to confirm

Reply:
Is the expected codegen then `mov, movprfx (predicated), ins (predicated)`?

It doesn't, in general, seem possible to always emit this as a 2-instruction sequence if using `movprfx` and there are several scenarios that require 3-instruction sequences instead. `movprfx` is documented to be a hint and may get optimized while `mov` is not.

If instead this were `ins (predicated), sel (predicated)` then it could always be a 2 instruction sequence (and in many cases be a 1 instruction sequence), but that is similarly not documented to be a hint or potentially optimized sequence.

Reply:
> The instruction after a movprfx can't be another movprfx.

Correct, these were the rules I implemented in binutils years ago, some of them aren't applicable to the jit of course:

* MOVPRFX must be followed by an instruction.
* MOVPRFX can only be followed by non-layout altering directives.
* MOVPRFX destination register MUST be used as the destination register in the
  instruction at PC+4, and is not allowed to be used in any other position other than
  destructive input.  This includes registers that architecturally overlap. e.g. x1
  should be treated as z1.
* MOVPRFX must be followed by a restricted set of SVE instructions.
* The size of the destination register of MOVPRFX must be equal to that of
  the operation at PC+4.
* The predicate register and operation of MOVPRFX must match that of the instruction
  at PC+4
* The predicated instruction at PC+4 must use the merging predicate.
* Architectural aliases and pseudo-instructions need to be supported as well.
* MOVPRFX cannot be the last instruction in a sequence

Reply:
> Is the expected codegen then mov, movprfx (predicated), ins (predicated)?
I this for the case of `ConditionalSelect(mask, Add(x,y), b)` where `b!=x` && `b!=y`?

I guess in these case we'd normally predicate the loads if single use.  But I could be missing some context here as I haven't read the entire thing, but why not do the operation and then use a sel (flipping the operands to flip the condition on the same predicate)

so for the above:
```
        add     z0.h, z0.h, z1.h
        sel     z0.h, p1, z2.h, z0.h
```
or did I miss something?

Reply:
`movprfx` is explicitly documented that hardware may take advantage of it and "fuse" the operation so that the otherwise destructive (RMW) operation can be treated as non-destructive.

There is no such similar documentation around `sel` even though it ultimately could be recognized and similarly pattern matched if hardware opted to do so.

------------------

So the question then is what is preferred from an architectural standpoint?

We can emit `mov, movprfx, ins` in which case the register renamer might be able to elide the `mov` and the backend may be able to fuse the `movprfx, ins` to be non-destructive. But, this takes 3 instructions (12-bytes) to encode.

Alternatively we can emit `ins, sel` and encode the same logical operation in 2 instructions (8-bytes) but there may not be any backend optimizations around this sequence.

We would then always prefer `movprfx, ins` for any sequences that can be represented using this sequence.

Reply:
> movprfx is explicitly documented that hardware may take advantage of it and "fuse" the operation so that the otherwise destructive (RMW) operation can be treated as non-destructive.

Sure, but architectural desires don't always line up with micro-architectural implementations.  As it stands today there are very few Arm micro-architectures that eliminate `SVE mov` at rename (as documented in their Software Optimization Guides).  And the situations in which `movprfx` can be eliminated by the micro-architecture is a bit constrained (also documented in the different SWoGs).

Running this sequence over dozens of different micro-architectures (both current and still in development) the `SEL` version always wins, even on those micro-architectures that can eliminate both the `mov` and the `movprfx`.  In practice they can't eliminate both when in a sequence as the above.

> We would then always prefer movprfx, ins for any sequences that can be represented using this sequence

yeah, whenever possible this should be done, we agree here.  It just doesn't seem wise to go out of the way to get a `movprfx` here.  At the very least most of the cores available at release time for the SVE support won't like it.

Reply:
@kunalspathak, this can basically be summed up as:
* where possible, emit just `ins`
* otherwise, prefer a 2 instruction sequence where `movprfx; ins` is better than `ins; sel`.

Reply:
> @kunalspathak, this can basically be summed up as:
> 
> * where possible, emit just `ins`

This will be done for "unpredicated" instruction

> * otherwise, prefer a 2 instruction sequence where `movprfx; ins` is better than `ins; sel` and `ins; sel` is better than a 3-instruction sequence.

1. `movprfx; ins` : This will be done for "unpredicated" RMW instruction where `targetReg != op1Reg`.
2. `ins; sel` : For instructions like `add` which has both predicated and unpredicated version, what we are saying is just do the "unpredicated" operation first and then use `sel` to select the active elements from the result. I am wondering if it will be ok to do this for "faulting" instructions where currently in the PR we are doing "ins" on just active elements, but with the new proposal, we would perform "ins" on all the lanes and then use `sel` to select just from active lanes. BTW, @TamarChristinaArm , i didn't understand your comment about " then use a sel (flipping the operands to flip the condition on the same predicate)". Can you please elaborate why this is not `sel     z0.h, p1, z0.h, z2.h` ?
3. 3-instruction sequence : Got it.

Reply:
> ins; sel : For instructions like add which has both predicated and unpredicated version, what we are saying is just do the "unpredicated" operation first and then use sel to select the active elements from the result. I am wondering if it will be ok to do this for "faulting" instructions where currently in the PR we are doing "ins" on just active elements, but with the new proposal, we would perform "ins" on all the lanes and then use sel to select just from active lanes. BTW, @TamarChristinaArm , i didn't understand your comment about " then use a sel (flipping the operands to flip the condition on the same predicate)". Can you please elaborate why this is not sel     z0.h, p1, z0.h, z2.h ?

No, this is rather for `ins (predicated); sel`. Not `ins (unpredicated); sel` which should've instead been emitted as `ins (predicated)` or `movprfx; ins (predicated)`.

----------------

In general we have two scenarios.

### Predication Not Required

When predication is not required, we want to emit:
1. ins (unpredicated)
2. ins (predicated)

The first is the simplest form, it's how we already emit all AdvSIMD instructions.

The second should only be done for instructions, like `abs (predicated)`, where no unpredicated version exists. They should always be 1 instruction and the mask used should be generated from `ptrue` (ideally reused from an early `ptrue`)

### Predication Required

When predication is required, we want to emit:
1. `ins (predicated)`
2. `movprfx (unpredicated); ins (predicated)`
4. `movprfx (predicated); ins (predicated)`
5. `ins (predicated); sel (predicated)`

Which we pick depends entirely on whether or not the instruction is RMW and what the value we're merging with is (zero, one of the inputs, or an independent value).

The first is the simplest form, but can only be used in special scenarios, such as:
* the instruction is not-RMW (`abs (predicated)`) and is being merged with `zero`
* the instruction is not-RMW and is being merged with the target register (`tgtReg == op3Reg`)
* the instruction is RMW and the destination is the same as the first input and we're merging with zero  (`tgtReg == op1Reg`)
* the instruction is RMW and the destination is the same as the first input and we're merging with the first input (`tgtReg == op1Reg == op2Reg`)

The second is a bit more complex and is basically the standard `RMW` pattern when `op1Reg != tgtReg` where you first do `mov tgtReg, op1Reg; ins tgtReg, op2Reg`. It exists for the cases like `abs (predicated)` when `tgtReg != op3Reg`, since we the merge operation is done with the destination.

The third is likewise more complex and used in similar `RMW` scenarios, primarily for cases like `add (predicated)` since when we're not merging with the first input. For example, if `tgtReg != op1Reg` then we first need to do a `movprfx (predicated)` to ensure `tgtReg` is `op1Reg` mixed with `op3Reg` (the merge value), we then need to do the `add (predicated)` to ensure that `tgtReg += op2Reg` is done but only for the values that were meant to be handled by the mask.

The fourth is because there are complex cases where due to register allocator or other limitations, there are some RMW instructions that would require `mov; movprfx (predicated); ins (predicated)`. These cases are ones where we'd instead prefer to emit `ins (predicated); sel (predicated)` to keep it a 2 instruction sequence. These should be generally rare

Reply:
Some examples, I think, would be:
```asm
; 1. op1 = cndsel(mask, op1 + op2, zero)
    movprfx op1, mask/Z, op1
    add     op1, mask/M, op2

; 2. op1 = cndsel(mask, op1 + op2, op1)
    add     op1, mask/M, op2

; 3. op1 = cndsel(mask, op1 + op2, op3)
    movprfx op1, mask/M, op3
    add     op1, mask/M, op2

; 4. dst = cndsel(mask, op1 + op2, zero)
    movprfx dst, mask/Z, op1
    add     dst, mask/M, op2

; 5. dst = cndsel(mask, op1 + op2, op1)
    movprfx dst, op1
    add     dst, mask/M, op2

; 6. dst = cndsel(mask, op1 + op2, op3)
    add     dst, op1, op2
    sel     dst, mask, dst, op3
```

For 1 we need 3 registers + zero. `op1` is both the input and output and we're merging with zero, so we can simply zero out the unused elements from `op1` with `movprfx (predicated)` and then do the `add (predicated)` to only update the entries we want to add.

For 2 we need 3 registers. `op1` is the input, output, and merge result. So we can simply emit `add (predicated)` only updating the entries we want to add.

For 3 we need 4 registers + zero. `op1` is the input and output, but we're merging with `op3`. In this case we need `movprfx (predicated)` to merge op1/op3 and then `add (predicated)` to only update the entries that we want to add.

For 4 we need 4 registers. We no longer have a register that is the input and output, but we're merging with zero. This means we can use `mov (predicated)` to clear the zeroed results in the destination, otherwise taking `op1`, and then do `add (predicated)` to only update the entries we want to add.

For 5 we need 4 registers. It's similar, but since we're merging with `op1` we can use `mov (unpredicated)` instead.

For 6 we need 5 registers. We then have the most complex case and we can't emit this trivially using the `add (predicated)` since that requires we first get `op3` into `dst` and then merge `dst` with `op1` before doing the `add (predicated)`. So, instead we fallback to doing a standard `add (unpredicated)` followed by a `sel (vectors)`.

Reply:
There are then some other optimization patterns that may be possible for cases like `op3 = cndsel(mask, op1 + op2, op3)` or similar, but they should follow the same general pattern of 1 or 2 instructions

Reply:
Thanks for the examples. I was actually in the process of preparing various combinations in the excel sheet and the expected instructions and will paste here shortly.

Reply:
> ; 3. op1 = cndsel(mask, op1 + op2, op3)
    movprfx op1, mask/M, op3
    add     op1, mask/M, op2

> For 3 we need 4 registers + zero. op1 is the input and output, but we're merging with op3. In this case we need movprfx (predicated) to merge op1/op3 and then add (predicated) to only update the entries that we want to add.

I think the `movprfx` needs to move only the lanes of `op3` that are inactive in `mask` so that it doesn't override the contents of active lane contents of `op1` that we have not yet added in `op2`. Here is the test I ran into:

Below, we just wipe the contents of `op1` present in `v16` using `movprfx` on `p7/m`. 

- target = v16
- op1 = v16
- op2 = v17
- falseOp = v18

```
            movprfx z16.s, p7/m, z18.s
            add     z16.s, p7/m, z16.s, z17.s
```

Instead it should do something like, which takes 3 instructions.

```asm
    not mask1, allbitsetMask, mask
    movprfx op1, mask1/M, op3
    add     op1, mask/M, op2
```

Do you remember the reasoning on why we just can't do:

```asm
    add     dst, op1, op2
    sel     dst, mask, dst, op3
```

so for the above example, it would be:

```asm
    add     z16.s, z16.s, z17.s
    sel     z16.s, mask, z16.s, z18.s
```

Reply:
> I think the movprfx needs to move only the lanes of op3 that are inactive in mask so that it doesn't override the contents of active lane contents of op1 that we have not yet added in op2. Here is the test I ran into:

Ah, you're right. I mixed up the operation in my head. `movprfx reg1, mask/M, reg2` operates as `reg1[i] = mask[i] ? reg2[i] : reg1[i]`, so it doesn't do what's needed and we likely need `add, sel` here as well.

@TamarChristinaArm, is there not a better way to do this here? We can certainly generate `add, sel`; but it seems better if we can emit the predicated versions of instructions instead.

Reply:
> > ; 3. op1 = cndsel(mask, op1 + op2, op3)
> > movprfx op1, mask/M, op3
> > add     op1, mask/M, op2

> Instead it should do something like, which takes 3 instructions.
> 
> ```assembly
>     not mask1, allbitsetMask, mask
>     movprfx op1, mask1/M, op3
>     add     op1, mask/M, op2
> ```
> 

That's not allowed, remember rule 6 from the list I posted, the predicates of the `movprfx` and the instruction need to match.

> > I think the movprfx needs to move only the lanes of op3 that are inactive in mask so that it doesn't override the contents of active lane contents of op1 that we have not yet added in op2. Here is the test I ran into:
> 
> Ah, you're right. I mixed up the operation in my head. `movprfx reg1, mask/M, reg2` operates as `reg1[i] = mask[i] ? reg2[i] : reg1[i]`, so it doesn't do what's needed and we likely need `add, sel` here as well.
> 
> @TamarChristinaArm, is there not a better way to do this here? We can certainly generate `add, sel`; but it seems better if we can emit the predicated versions of instructions instead.

Unfortunately not, for this to work you'd have needed a destructive three operand predicated add where destination and input didn't need to be the same. So I think add + sel is the only way, 

You might be able to peephole it away later to a `movprfx` if the next instruction supports it though.

Reply:
> Since this is an optimization and other API implementation is dependent on this PR, I want to go ahead and merge this PR (if nothing else stands out) and do the follow-up work, if needed.

+1. And add a new issue for tracking it.